### ISSUE TITLE
Process finished with exit code 139 (interrupted by signal 11: SIGSEGV)

### ISSUE THREAD

**Issue by ghost** [uWebSockets/uWebSockets#474]
Ubuntu Gnome 16.10 64bit here. The server crashes all the time. I'm using `master` branch. Not using SSL.

```bash
$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/6/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 6.2.0-5ubuntu12' --with-bugurl=file:///usr/share/doc/gcc-6/README.Bugs --enable-languages=c,ada,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-6 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-libmpx --enable-plugin --enable-default-pie --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-6-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-6-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-6-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 6.2.0 20161005 (Ubuntu 6.2.0-5ubuntu12) 
```

I want to check the Origin header in the server side. For example, if it isn't `http://localhost` it should close the connection. There's a function called `onHttpUpgrade`, but if I add `onHttpUpgrade` handler, it stops upgrading clients. It would be great, if onHttpUpgrade was bool, for example, if you return `true` the websocket is open, but if you return `false` the websocket is closing.

I tried to make a workaround (`tests` is the name of the project):

 main.cpp
```c++
#include <uWS/uWS.h>

int main() {
    uWS::Hub h;

    h.onConnection([](uWS::WebSocket<uWS::SERVER> ws, uWS::HttpRequest handshake) {
        ws.close(); // skipped the origin filtering
    });

    h.listen(3000);
    h.run();

    return 0;
}
```

CMakeLists.txt:
```cmake
cmake_minimum_required(VERSION 3.6)
project(tests)

set(CMAKE_CXX_STANDARD 11)

set(SOURCE_FILES main.cpp)
add_executable(tests ${SOURCE_FILES})
target_link_libraries(tests uWS)
target_link_libraries(tests z)
```

This is all what I have.

Every time I run this Node.JS code (this is whole file too):
```javascript
var WebSocket = require('ws');
var clients = [];

for (var i = 0; i < 1000; i++) {
	var ws = new WebSocket("ws://localhost:3000");
	clients.push(ws);
}
```

I get `Process finished with exit code 139 (interrupted by signal 11: SIGSEGV)`.
Also I get [sometimes]
```
tests: src/unix/poll.c:95: uv_poll_stop: Assertion `!uv__is_closing(handle)' failed.
```

**Comment by SzAmmi** [comment]
Related: https://github.com/uWebSockets/uWebSockets/issues/440

**Comment by SzAmmi** [comment]
@alexhultman Just updated the main post, I think the issue template is filled properly now.

**Comment by SzAmmi** [comment]
Sometimes I get uv poll error, but it stopped appearing.

**Comment by SzAmmi** [comment]
Btw, this label seems to be new :P

**Comment by alexhultman** [comment]
Yes it is for people who don't read what has been written

**Comment by alexhultman** [comment]
I cannot reproduce these problems. They seem too fundamental to be an issue in the library and chances are you have issues on your end. It is not completely unheard of that people actually compile with one version of a library and then link to another one only to get a mix that blows up in their face. Try replacing the C++ part with equivalent Node.js code and see if you get the same issues. If not then you need to debug your part and track down where it crashes and why.

**Comment by SzAmmi** [comment]
@alexhultman ok.

**Comment by SzAmmi** [comment]
IKR you think it's my fault, but maybe. I've just created a new project. It has only two files:

CMakeLists.txt:
```cmake
cmake_minimum_required(VERSION 3.6)
project(tests)

set(CMAKE_CXX_STANDARD 11)

find_library(LIBBOX2D Box2D)

set(SOURCE_FILES main.cpp)
add_executable(tests ${SOURCE_FILES})
target_link_libraries(tests uWS)
target_link_libraries(tests z)
```

main.cpp:
```c++
#include <uWS/uWS.h>

int main() {
    uWS::Hub h;

    h.onConnection([&](uWS::WebSocket<uWS::SERVER> ws, uWS::HttpRequest handshake) {
        std::string origin = handshake.getHeader("origin").toString();

        if (origin != "http://localhost" && origin != "http://cosmoc.io" && origin != "http://192.168.0.150") {
            ws.close();
        }
    });

    h.listen(3000);
    h.run();

    return 0;
}
```

Every time I run
```javascript
var WebSocket = require('ws');
var clients = [];

for (var i = 0; i < 1000; i++) {
	var ws = new WebSocket("ws://localhost:3000");
	clients.push(ws);
}
```
I get `Process finished with exit code 139 (interrupted by signal 11: SIGSEGV)`.

It happens on the Node.JS version too.
Server code:
```javascript
var WebSocketServer = require('uws').Server;
var wss = new WebSocketServer({ port: 3000 });

console.log("Server started");

wss.on('connection', function(ws) {
    wss.close(); // I was lazy to filter the origin header
});
```
The task just kills automatically after I run the file with 1k clients test.

**Comment by SzAmmi** [comment]
Node.JS client (client connets to the server, I'm using `ws` to connect to `uws`) output:
```javascript
$ node test
events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: socket hang up
    at createHangUpError (_http_client.js:254:15)
    at Socket.socketOnEnd (_http_client.js:346:23)
    at emitNone (events.js:91:20)
    at Socket.emit (events.js:185:7)
    at endReadableNT (_stream_readable.js:974:12)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9)

```

**Comment by SzAmmi** [comment]
Server output:
```bash
$ node uws
Server started
$ # the process was killed :|
```

**Comment by SzAmmi** [comment]
The socket shouldn't hang up on `ws.close`, it should just close normally.

**Comment by SzAmmi** [comment]
Even if I add `ws.onerror` to the client, the process kills itself, probably it's caused by SIGSEGV too.

**Comment by SzAmmi** [comment]
Oh wait, I just did wss.close instead of ws.close. My fault there, node.js version closes the connections without any problems. But I don't know why the C++ project fails. I'll post a link to zip package.

**Comment by alexhultman** [comment]
Right. Again, check your end thoroughly and stop spamming. Post next message tomorrow

**Comment by SzAmmi** [comment]
@alexhultman I've updated my main post. I explained everything there. Sorry, but I can't tell anything more. That's all I have to say.

**Comment by SzAmmi** [comment]
@alexhultman Just one more info: If I add
```c++
std::this_thread::sleep_for(std::chrono::seconds(1));
```
before `ws.close()` everything goes OK. I tried to run it on another thread, but it fails. So here's my result:

_if I try to close 2 (or more) connections opened "at the same time", the server fails._

**Comment by alexhultman** [comment]
You still haven't provided any debug output, any asan reports or anything. I hear your problem but I cannot reproduce it and I won't look at this until someone else also has this problem. You have a history of posting issues only to close them later on when you find that it was on your end. Don't spam this issue, debug it and fix it

**Comment by alexhultman** [comment]
Did you try with uws 0.13 in Node.js?

**Comment by SzAmmi** [comment]
```javascript
var WebSocketServer = require('uws').Server;
var wss = new WebSocketServer({ port: 3000 });

wss.on('connection', function(ws) {
    ws.close();
});
```
It works as expected - no errors here. But Node.JS is slow for real multiplayer game servers, so this is why I chose C++ and uWS. 

"You still haven't provided any debug output" - What I have to do? This is the simplest server I could make: close connection in the onConnection handler.

"any asan reports" - will do right now.

**Comment by SzAmmi** [comment]
@alexhultman Here's tha asan report (address only):

```
/home/szami/CLionProjects/tests/cmake-build-debug/tests
ASAN:DEADLYSIGNAL
=================================================================
==10436==ERROR: AddressSanitizer: SEGV on unknown address 0x60f0488000d2 (pc 0x7f4f3bc69e6c bp 0x7ffd2e5ce3e0 sp 0x7ffd2e5ce3d0 T0)
    #0 0x7f4f3bc69e6b in uWS::Group<true>::addHttpSocket(uv_poll_s*)::{lambda(uv_timer_s*)#1}::operator()(uv_timer_s*) const::{lambda(uWS::HttpSocket<true>)#1}::operator()(uWS::HttpSocket) const (/usr/lib/libuWS.so+0x3ae6b)
    #1 0x7f4f3bc6bb8f in void uWS::Group<true>::forEachHttpSocket<uWS::Group<true>::addHttpSocket(uv_poll_s*)::{lambda(uv_timer_s*)#1}::operator()(uv_timer_s*) const::{lambda(uWS::HttpSocket<true>)#1}>(uWS::Group<true>::addHttpSocket(uv_poll_s*)::{lambda(uv_timer_s*)#1}::operator()(uv_timer_s*) const::{lambda(uWS::HttpSocket<true>)#1} const&) (/usr/lib/libuWS.so+0x3cb8f)
    #2 0x7f4f3bc69e38 in uWS::Group<true>::addHttpSocket(uv_poll_s*)::{lambda(uv_timer_s*)#1}::operator()(uv_timer_s*) const (/usr/lib/libuWS.so+0x3ae38)
    #3 0x7f4f3bc69ef3 in uWS::Group<true>::addHttpSocket(uv_poll_s*)::{lambda(uv_timer_s*)#1}::_FUN(uv_timer_s*) (/usr/lib/libuWS.so+0x3aef3)
    #4 0x7f4f39855a27 in uv__run_timers src/unix/timer.c:165
    #5 0x7f4f39849c3b in uv_run src/unix/core.c:343
    #6 0x7f4f3bc7dfc5 in uS::Node::run() (/usr/lib/libuWS.so+0x4efc5)
    #7 0x55a5b258bb52 in main /home/szammi/CLionProjects/tests/main.cpp:11
    #8 0x7f4f3a3b73f0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x203f0)
    #9 0x55a5b258b8c9 in _start (/home/szammi/CLionProjects/tests/cmake-build-debug/tests+0xc8c9)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV (/usr/lib/libuWS.so+0x3ae6b) in uWS::Group<true>::addHttpSocket(uv_poll_s*)::{lambda(uv_timer_s*)#1}::operator()(uv_timer_s*) const::{lambda(uWS::HttpSocket<true>)#1}::operator()(uWS::HttpSocket) const
==10436==ABORTING

Process finished with exit code 1
```

**Comment by 0x111111111** [comment]
This error seems to be fixed now (release v0.13.0), for me any errors don't appear, everything's ok. The author of the issue probably has used old version of uWebSockets.

**Comment by alexhultman** [comment]
It never was an issue. I even have tests for this very thing. User most likely messed up
